File: javalin/src/main/java/io/javalin/core/plugin/PluginLifecycleViolationException.java
package io.javalin.core.plugin;

import org.jetbrains.annotations.NotNull;

import java.util.Objects;

public abstract class PluginLifecycleViolationException extends RuntimeException {

    private final Class<? extends Plugin> pluginClass;

    public PluginLifecycleViolationException(@NotNull Class<? extends Plugin> pluginClass, String message) {
        super("Error in " + pluginClass.getCanonicalName() + " " + message);
        this.pluginClass = pluginClass;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof PluginLifecycleViolationException)) return false;
        PluginLifecycleViolationException that = (PluginLifecycleViolationException) o;
        return pluginClass.equals(that.pluginClass);
    }

    @Override
    public int hashCode() {
        return Objects.hash(pluginClass);
    }
}

File: javalin/src/main/java/io/javalin/core/plugin/PluginLifecycleViolationException.kt
package io.javalin.core.plugin

abstract class PluginLifecycleViolationException(
        open val pluginClass: Class<out Plugin>,
        override val message: String
) : RuntimeException("Error in ${pluginClass.canonicalName}: $message")

data class PluginInitLifecycleViolationException(override val pluginClass: Class<out Plugin>) : PluginLifecycleViolationException(
        pluginClass,
        "It is not allowed to register handlers during the 'Init' lifecycle. Please use the 'Apply' lifecycle instead."
)
